platform/solr/solr-factory/src/main/java/org/codice/solr/client/solrj/SolrClient.java

```diff
@@ -0,0 +1,1030 @@
+
+  /**
+   * Gets the core associated with this Solr client
+   *
+   * @return the core associated with this Solr client
+   */
+  public String getCore();
+
+  /**
+   * Checks whether the Solr server this client is connected to and the corresponding core is
+   * available or not. This method is different from an active ping as it might be basing its answer
+   * on cached information. It might, however, initiate a background ping.
+   *
+   * <p><i>Note:</i> Returning <code>true</code> doesn't guarantee the next call to any methods will
+   * succeed as the server or the core might become unavailable at any time without notice.
+   *
+   * @return <code>false</code> if the Solr server this client is connected or the corresponding
+   *     core is not available; <code>true</code> otherwise
+   */
+  public boolean isAvailable();
+
+  /**
+   * Waits if necessary for the Solr server this client is connected to and the corresponding core
+   * to become available. This method is different from an active ping as it might be basing its
+   * answer on cached information. It might, however, initiate a background ping.
+   *
+   * <p><i>Note:</i> Nothing will happen and the method will return right away if the client was
+   * already closed.
+   *
+   * <p><i>Note:</i> Returning <code>true</code> doesn't guarantee the next call to any methods will
+   * succeed as the server or the core might become unavailable at any time without notice.
+   *
+   * @param timeout the maximum time to wait
+   * @param unit the time unit of the timeout argument
+   * @return <code>false</code> if the Solr server this client is connected or the corresponding
+   *     core is not available; <code>true</code> otherwise
+   * @throws InterruptedException if the current thread was interrupted while waiting
+   */
+  public boolean isAvailable(long timeout, TimeUnit unit) throws InterruptedException;
+
+  /**
+   * This method is the same as {@link #isAvailable()} but allows a caller to provide a listener
+   * that will be called back with a boolean flag representing the new availability for this Solr
+   * client whenever it changes in the future. The listener will also be called back right away with
+   * the current availability of the client. This method is different from an active ping as it
+   * might be basing its initial answer on cached information. It might, however, initiate a
+   * background ping.
+   *
+   * <p><i>Note:</i> Nothing will happen and the method will return right away if the client was
+   * already closed.
+   *
+   * <p><i>Note:</i> Returning <code>true</code> doesn't guarantee the next call to any methods will
+   * succeed as the server or the core might become unavailable at any time without notice.
+   *
+   * @param listener a listener to notify whenever the Solr client's availability changes
+   * @return <code>false</code> if the Solr server this client is connected or the corresponding
```

Comment:
❓Since we're calling the listener right away with the current state and the client will always implement the state change logic in the listener, is there real value in returning a `boolean`?

Reply:
I kept those interface identical to the catalog provider interfaces. Personally, I wasn't found of a isAvailable() with listener. In any case, it doesn't hurt anything as you discovered, the listener guarantees an initial call right away which in my experience does simplify using listeners as it removes the race condition window that typically exists in such interfaces.

I would be ok to remove it except for the fact that the catalog provider would have to be re-implemented in a more complex way as it does return a boolean.